### DISCUSSION TITLE
Support for elasticsearch 7.x

### DISCUSSION THREAD

**Discussion by bogdanlupu** [wallneradam/esorm#24] [Q&A] [answered]
Hi, I noticed you mentioned testing it with ES 7.x and 8.x, however in the requirements it is currently limited to >=8.6.0, <9.0.0. Since the ES version needs the corresponding elasticsearch package version, how is it possible to install elasticsearch 7.x and pyesorm together?

**Comment by wallneradam** [comment]
Hi,
Because that is the python client version, not ES version. It is indeed backward compatible, but 8.x ES features won't work. Almost all the features ESORM can do is supported and tested on 7.x ES version. You can check my test cases, it uses Docker to install and test ES 7.x and 8.x versions and test (almost) everything on both of them.

**Reply by bogdanlupu** [reply]
Thanks for the quick reply, was not expecting it!
The reason I was asking is because I saw this on their documentation
![image](https://github.com/user-attachments/assets/9bd07176-46a8-42c5-a7f7-43271833bddd)
And when I try to run stuff locally using ES 7 I get `'Content-Type header [application/vnd.elasticsearch+json; compatible-with=8] is not supported'`, which lead me to believe it could be due to a version mismatch.

**Reply by wallneradam** [reply]
It says: "Elasticsearch language clients are only backwards compatible with default distributions and without guarantees made."

I used only the latest version of 7.x on tests. Which version you use? Is it a warning only and works? Or it does not work?

Anyway I'm testing the v7.x python library and ESORM can work on top of it with some minor modifications, so I may change the requirement to be able to use lower version because of the support table you referred. I will make a version with that modification.

**Reply by bogdanlupu** [reply]
It does not work, I get a 406 back with the error message I pasted above. I'm on 7.10.1

**Comment by wallneradam** [comment]
It is now supported in v0.6.2 to use older (7.x) library versions. 
Could you please check it?

**Reply by bogdanlupu** [reply]
Connection works now, but I get an error when trying to `setup_mappings`:
`E       TypeError: IndicesClient.create() got an unexpected keyword argument 'mappings'`
Apparently it stems from the fact that the signature for `create` in 7.x is `async def create(self, index, body=None, params=None, headers=None)`, but it is called with
```
es.indices.create(index=index,
                           mappings={'properties': properties},
                           settings=model.ESConfig.settings,
                           request_timeout=90)
```

**Reply by wallneradam** [reply, marked as answer]
Hmm, which version of the library you use? I've tested with the newest 7.17.12, which has this:
```python
@query_params(
        "include_type_name",
        "master_timeout",
        "timeout",
        "wait_for_active_shards",
        request_mimetypes=["application/json"],
        response_mimetypes=["application/json"],
        body_params=["aliases", "mappings", "settings"],
)
```
So it is supported with both mappings and body.

You don't need to use the same version of the library as the ES itself. If you think it is a problem, I can create a try catch around it to make the older syntax also possible. Or pull request is welcome. But I think you should just update to the latest 7.x version of the lib.

**Reply by bogdanlupu** [reply]
Can confirm it works on 7.17.x, previously I was on 7.10.x, was expecting it to be backwards compatible across a given major version but it seems that's not the case. Might be worth mentioning as a caveat, or limiting the ES library version to 7.17. 
I don't think it's worth adding the complexity to support <7.17, as it does not really make a difference for the user, presumably.
Thank you for the prompt replies and fix!

**Reply by wallneradam** [reply]
It will be limited to >=7.17 in the next release.
Thanks for the feedback.
